### PR TITLE
Support customization of numOfProbe and probeInterval when externaltrafficpolicy is local

### PR DESCRIPTION
**Pull request by MartinForReal**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Support customization of numOfProbe and probeInterval when externaltrafficpolicy is local
#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:
move health probe related code to azure_loadbalancer_healthprobe*
#### Does this PR introduce a user-facing change?

```release-note
Support customization of numOfProbe and probeInterval when externaltrafficpolicy is local
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/provider/azure_loadbalancer_healthprobe.go [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+				if strings.EqualFold(item.Name, *probePort) {
+					//found the port
+					properties.Port = pointer.Int32(item.NodePort)
+				}
+			}
+		} else {
+			// Not need to verify probePort is in correct range again.
+			for _, item := range serviceManifest.Spec.Ports {
+				//nolint:gosec
+				if item.Port == int32(port) {
```
[the remark this answers is not in the document]
Comment (MartinForReal) [reply pkg/provider/azure_loadbalancer_healthprobe.go:216]: upper bound check has been added. false alarm.

**Comment by MartinForReal** [comment]
<html>
<body>


/test pull-cloud-provider-azure-e2e-ccm-capz
--


<br class="Apple-interchange-newline">
</body>
</html>

File: pkg/consts/consts.go [hunk 2]
```diff
@@ -442,47 +442,6 @@ const RateLimited = "rate limited"
 // CreatedByTag tag key for CSI drivers
 const CreatedByTag = "k8s-azure-created-by"
 
-// port specific
-const (
```
Comment (feiskyer) [inline pkg/consts/consts.go:446]: Could we keep them in the consts package (as they are shared by both provider and tests)?

File: tests/e2e/network/service_annotations.go [hunk 3]
```diff
@@ -772,12 +773,12 @@ var _ = Describe("Service with annotation", Label(utils.TestSuiteLabelServiceAnn
...
 		By("Creating a service with health probe annotations")
 		annotation := map[string]string{
-			consts.ServiceAnnotationLoadBalancerHealthProbeNumOfProbe:                                      "5",
-			consts.BuildHealthProbeAnnotationKeyForPort(serverPort, consts.HealthProbeParamsNumOfProbe):    "3",
-			consts.ServiceAnnotationLoadBalancerHealthProbeInterval:                                        "15",
-			consts.BuildHealthProbeAnnotationKeyForPort(serverPort, consts.HealthProbeParamsProbeInterval): "10",
-			consts.ServiceAnnotationLoadBalancerHealthProbeProtocol:                                        "Http",
-			consts.ServiceAnnotationLoadBalancerHealthProbeRequestPath:                                     "/",
+			consts.ServiceAnnotationLoadBalancerHealthProbeNumOfProbe:                                          "5",
+			provider.BuildHealthProbeAnnotationKeyForPort(serverPort, provider.HealthProbeParamsNumOfProbe):    "3",
```
Comment (feiskyer) [inline tests/e2e/network/service_annotations.go:777]: e2e should not import provider directly

File: pkg/consts/consts.go (hunk 2 above)
Comment (MartinForReal) [reply pkg/consts/consts.go]: the change is reverted.

File: tests/e2e/network/service_annotations.go (hunk 3 above)
Comment (MartinForReal) [reply tests/e2e/network/service_annotations.go]: the change is reverted.

File: pkg/provider/azure_loadbalancer_healthprobe.go [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+	}
+	probe := &network.Probe{
+		Name:                  &lbrule,
+		ProbePropertiesFormat: properties,
+	}
+	return probe, nil
+}
+
+// getHealthProbeConfigProbeIntervalAndNumOfProbe
+func (az *Cloud) getHealthProbeConfigProbeIntervalAndNumOfProbe(serviceManifest *v1.Service, port int32) (*int32, *int32, error) {
```
Comment (MartinForReal) [inline pkg/provider/azure_loadbalancer_healthprobe.go:183]: newly added.

File: pkg/provider/azure_loadbalancer_healthprobe.go [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+	if (*probeInterval)*(*numberOfProbes) >= 120 {
+		return nil, nil, fmt.Errorf("total probe should be less than 120, please adjust interval and number of probe accordingly")
+	}
+	return probeInterval, numberOfProbes, nil
+}
+
+// getHealthProbeConfigProbeInterval get probe interval in seconds
+// minimum probe interval in seconds is 5. ref: https://docs.microsoft.com/en-us/rest/api/load-balancer/load-balancers/create-or-update#probe
+// if probeInterval is not set, set it to default instead ref: https://docs.microsoft.com/en-us/rest/api/load-balancer/load-balancers/create-or-update#probe
+func (*Cloud) getHealthProbeConfigProbeInterval(serviceManifest *v1.Service, port int32) (*int32, error) {
```
Comment (MartinForReal) [inline pkg/provider/azure_loadbalancer_healthprobe.go:204]: newly added.

File: pkg/provider/azure_loadbalancer_healthprobe.go [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+	if probeInterval == nil {
+		probeInterval = pointer.Int32(consts.HealthProbeDefaultProbeInterval)
+	}
+	return probeInterval, nil
+}
+
+// getHealthProbeConfigNumOfProbe get number of probes
+// minimum number of unhealthy responses is 2. ref: https://docs.microsoft.com/en-us/rest/api/load-balancer/load-balancers/create-or-update#probe
+// if numberOfProbes is not set, set it to default instead ref: https://docs.microsoft.com/en-us/rest/api/load-balancer/load-balancers/create-or-update#probe
+func (*Cloud) getHealthProbeConfigNumOfProbe(serviceManifest *v1.Service, port int32) (*int32, error) {
```
Comment (MartinForReal) [inline pkg/provider/azure_loadbalancer_healthprobe.go:233]: newly added.

File: pkg/provider/azure_loadbalancer_healthprobe.go [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+				return nil, fmt.Errorf("failed to parse annotation %s: %w", consts.ServiceAnnotationLoadBalancerHealthProbeRequestPath, err)
+			}
+		}
+		if path == nil {
+			path = pointer.String(consts.HealthProbeDefaultRequestPath)
+		}
+		properties.RequestPath = path
+	}
+
+	properties.IntervalInSeconds, properties.ProbeThreshold, err = az.getHealthProbeConfigProbeIntervalAndNumOfProbe(serviceManifest, port.Port)
```
Comment (MartinForReal) [inline pkg/provider/azure_loadbalancer_healthprobe.go:171]: extracted three functions from here.

File: pkg/provider/azure_loadbalancer.go [hunk 8]
```diff
@@ -2396,15 +2193,18 @@ func (az *Cloud) getExpectedLBRules(
 	if servicehelpers.NeedsHealthCheck(service) && !(consts.IsPLSEnabled(service.Annotations) && consts.IsPLSProxyProtocolEnabled(service.Annotations)) {
 		podPresencePath, podPresencePort := servicehelpers.GetServiceHealthCheckPathPort(service)
 		lbRuleName := az.getLoadBalancerRuleName(service, v1.ProtocolTCP, podPresencePort, isIPv6)
-
+		probeInterval, numberOfProbes, err := az.getHealthProbeConfigProbeIntervalAndNumOfProbe(service, podPresencePort)
```
Comment (MartinForReal) [inline pkg/provider/azure_loadbalancer.go:2196]: newly added.

**Review by feiskyer — COMMENTED** [review]
/lgtm

**Comment by nimimeht** [comment]
@MartinForReal - was this PR successfully cherry-picked for release-1.27? I don't see the merge commit for this PR in the release changelog: https://github.com/kubernetes-sigs/cloud-provider-azure/compare/v1.27.5...v1.27.6

**Comment by MartinForReal** [comment]
@nimimeht please refer to https://github.com/kubernetes-sigs/cloud-provider-azure/commit/48ccfc3ddfd300058702d757bb85f27c55b7e545

**Pull request merged**